### PR TITLE
Footnotes initial support

### PR DESCRIPTION
**Pull request by git-noise**
## What does this change?
Add initial support for footnotes. Attempts to fix https://github.com/bokuweb/docx-rs/issues/716.
This focuses on the rust-side of the code - i.e docx-core - but snapshots for docx-wasm have been updated so that tests pass.

## References

- See https://github.com/bokuweb/docx-rs/issues/716 for details.

## Screenshots

Not applicable

## What can I check for bug fixes?

An example has been created at `docx-core/examples/footnotes.rs`.
```sh
cargo run --example footnotes
```

### CODE REVIEW

**Comment by bokuweb** [comment]
@git-noise Thanks for your great work! I'll check it later :)

**Comment by git-noise** [comment]
I fixed the clippy issue, I was using a more recent version that was flagging more lints, and this one got lost in the middle.
I ran clippy 1.56 and it passed locally, so it should hopefully pass the workflow now.

File: docx-core/examples/footnotes.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    Docx::new()
+        .add_paragraph(
+            Paragraph::new()
+                .add_run(Run::new().add_text("Hello"))
+                .add_run(
+                    Run::new().add_footnote_reference(
+                        Footnote::new(1)
+                            .add_content(Paragraph::new().add_run(Run::new().add_text("World"))),
+                    ),
+                ),
```
Comment (bokuweb) [inline docx-core/examples/footnotes.rs:15]: @git-noise 
Thanks for your great work!!!!!!

[question]
In `docx-rs`, I'm using an interface extensively that automatically assigns IDs. Can this feature be implemented to automatically assign IDs as well, in the following interface?"

``` rust
Run::new().add_footnote(
  Footnote::new().add_content(Paragraph::new().add_run(Run::new().add_text("World"))),
)
```

ref:  https://github.com/bokuweb/docx-rs/blob/main/docx-core/src/documents/hyperlink_id.rs

Comment (git-noise) [reply docx-core/examples/footnotes.rs]: Hello,

I did the modification, for now only on `Footnotes` themselves not `FootnoteReference`.

The reasoning is that in the current code, while one can technically instantiate a `FootnoteReference` by directly calling `FootnoteReference::new()`, the intended usage - also showcased in the example you mentioned above - is to use `add_footnote_reference()` on a `Footnote`.
Behind the scene, a `FootnoteReference` is then created from `Footnotes` by calling `into()`.

If we want to also add automatic IDs for FootnoteReference themselves, then we may hit a synchronization issue, as we may need to keep the FootnoteReference ID and Footnote ID in sync.

The current setup probably strikes a balance between usability and simplicity, but let me know your thoughts.

Comment (bokuweb) [reply docx-core/examples/footnotes.rs]: @git-noise 

Thank you. In cases where the user does not need to be aware of the id, I want the id to be hidden and added automatically. Other functions in this library are designed this way. Therefore, if possible, I would like the id to be auto-generated.

Comment (git-noise) [reply docx-core/examples/footnotes.rs]: I think it currently is the case. With the latest modification, a user would call something like the following:
```rust
Run::new().add_footnote_reference(
  Footnote::new().add_content(
    Paragraph::new().add_run(Run::new().add_text("World"))
  )
)
```
In this case there is no need to give an ID, and behind the scene, the `Footnote` and `FootnoteReference` are created with the same ID.

Comment (bokuweb) [reply docx-core/examples/footnotes.rs]: @git-noise Sorry! thanks for your work :)

**Review by bokuweb — APPROVED** [review]
@git-noise LGTM!!!!!!!

**Pull request merged**